### PR TITLE
Rho ADT and Monadic Interpreter

### PR DESCRIPTION
**Pull request by JosDenmark**
ADT/Proc.scala contains an ADT for constructing rho-calculus expressions. The ADT is extended with the `New` expression for generating fresh channels. In addition to the ADT, instances of `Functor[Proc]`, `Foldable[Proc]`, and `Traverse[Proc]` are implemented for a basic container API.

RhoInterface defines a monadic abstract machine for the rho-calculus. The state of the rho-machine consists of a store and a run-queue. The store is a mapping of channels to the processes waiting to read or write on those channels (`ChannelQueue`). The run-queue is simply a list of processes waiting to be evaluated. The helper functions defined in the interface manipulate the machine state in some way.

The rho-machine evaluates a rho-calculus expression for every possible interleaving. Inter-leavings are explored by permuting the run-queue. In particular, any process `P | Q` has two possible interleavings: 1) first `P`, then `Q`, and 2) first `Q`, then `P`. For each possible interleaving, the rho-machine will produce a new run-queue, a new machine state, and continue evaluating, returning an execution trace for each permutation.

The logic for evaluation, trace logging, and inter-leaving exploration is wrapped in the `Trace` monad transformer - the composition `StateT[WriterT[List, W, ?]], S, A]`. Given an initial state, the trace transformer returns a list of possible execution traces, `List[S]`, a final states, `S`, and an output value, `A`.

### CODE REVIEW

**Comment by nashef** [comment]
Please give your pull request a meaningful title and a thorough description of its contents.

File: rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+            //@*@Q = @Q in P{@*@Q/z}
+            if (name.equals(chan0)) cancel(chan1)
+            else name
+          }
+
+
+  val write: Channel => ChannelQueue => Trace[MachineState, Unit] =
+    chan =>
+      chanQ =>
+        for {store <- getStore
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala]: Looks wrong to me to have the `for` and `store <- getStore` in one line.

File: rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+
+  case class ReaderQueue(x:Reader, xs: List[Reader]) extends ChannelQueue {
+    override def toString: String = (x :: xs).map{_.toString}.mkString("[", "][", "]")
+  }
+
+  case class WriterQueue(x: Writer, xs: List[Writer]) extends ChannelQueue {
+    override def toString: String = (x :: xs).map{_.toString}.mkString("[", "][", "]")
+  }
+
+  case class EmptyQueue() extends ChannelQueue {
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala]: `case object`?

File: rholang/src/main/scala/AbstractInterpreter/Trace.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      for { (a, state1, log0) <- reduce(state0)
+            (b, state2, log1) <- f(a).reduce(state1) } yield {
+        (b, state2, log0 ++ log1)
+      }
+  }
+}
+
+object Trace {
+
+  def get[S]: Trace[S, S] = {
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/Trace.scala]: Unnecessary brackets.

File: rholang/src/main/scala/AbstractInterpreter/Trace.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    state { st => (st, st) }
+  }
+
+  def put[S]: S => Trace[S, Unit] = {
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/Trace.scala]: Unnecessary brackets.

File: rholang/src/main/scala/AbstractInterpreter/Trace.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
+package AbstractInterpreter
+
+import cats.Monad
+
+//Trace := StateT[WriterT[List,List,(S,A)]] = S => WriterT[List,List,(S,A)]] = S => (S,A) => List[(List[S],(S,A)] = List[(A,S,List[S])]
+
+//Trace is a monad transformer that takes a state and returns a list of possible end-states,
+//a value, A, and, for each end-state, a list of intermediate states.
+
+case class Trace[S,A](reduce: S => List[(A, S, List[S])]) {
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/Trace.scala]: Hmm, this should be a typeclass I guess?

File: rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+
+
+  val write: Channel => ChannelQueue => Trace[MachineState, Unit] =
+    chan =>
+      chanQ =>
+        for {store <- getStore
+             _ <- putStore(store + {
+               chan -> chanQ
+             })
+        } yield ()
```
Comment (birchmd) [inline rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala]: I don't think the `yield` keyword or brackets are necessary when you're not returning anything

File: rholang/src/main/scala/ADT/Proc.scala [hunk 7]
```diff
@@ -0,0 +1,118 @@
+package ADT
+
+import cats.{Applicative, Eval, Foldable, Functor, Traverse}
+
+// Term constructors
+trait Proc[+Chan] extends Serializable {
```
Comment (tymm) [inline rholang/src/main/scala/ADT/Proc.scala]: This should probably be a `sealed trait`.

File: rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+  case class Var(id: String) extends Channel {
+    override def toString: String = id
+  }
+
+//A ChannelQueue may be an empty queue, a list of readers, or a list of writers.
+// It will never be both a list of readers and a list of writers
+sealed trait ChannelQueue
+
+  case class ReaderQueue(x:Reader, xs: List[Reader]) extends ChannelQueue {
+    override def toString: String = (x :: xs).map{_.toString}.mkString("[", "][", "]")
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala]: `.map(_.toString)`.

File: rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+  }
+
+  case class WriterQueue(x: Writer, xs: List[Writer]) extends ChannelQueue {
+    override def toString: String = (x :: xs).map{_.toString}.mkString("[", "][", "]")
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala]: `.map(_.toString)`.

File: rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+  }
+
+  case class EmptyQueue() extends ChannelQueue {
+    override def toString: String = "[]"
+  }
+
+//A reader is a abstraction providing a bound name "z", and a continuation
```
Comment (tymm) [inline rholang/src/main/scala/AbstractInterpreter/RhoInterface.scala]: All these outer multi-line comments should be in this format:
```
/**
  * A reader is [...]
  */
````
If it is just one line then: `/** A reader is [...] */`

File: rholang/src/main/scala/AbstractInterpreter/Trace.scala (hunk 5 above)
Comment (DoSOfRedRiver) [reply rholang/src/main/scala/AbstractInterpreter/Trace.scala]: It doesn't look like a transformer, but a concrete monad. 
Why you don't just use `StateT[WriterT[List,List,(S,A)]]` from `cats`?

Comment (JosDenmark) [reply rholang/src/main/scala/AbstractInterpreter/Trace.scala]: This pull request is four days old. In the up to date version `type Trace[S,W,A] = StateT[WriterT[List,W,(S,A)]]`

File: rholang/src/main/scala/interpreter/Proc.scala [hunk 11]
```diff
@@ -159,6 +163,10 @@ object Proc {
     }
   }
 
+  /*
```
Comment (tymm) [inline rholang/src/main/scala/interpreter/Proc.scala]: Use this style for multi-line comments:
```
 /**
  * Differs from [...]
  * cancels to n.
  */
```

See:  https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards

File: rholang/src/main/scala/interpreter/Proc.scala [hunk 12]
```diff
@@ -320,12 +329,12 @@ object Proc {
       case Nil => (None, procList)
       case hProc :: rProc =>
         proc match {
-          case Zero    => (None, procList) //(* 0 has no match *)
-          case Drop(n) => (None, procList) //(* Drop has no match *)
+          case Zero    => (None, procList) //* 0 has no match */
```
Comment (tymm) [inline rholang/src/main/scala/interpreter/Proc.scala]: Hmm, never saw `//* [...] */` before. I would use `// 0 has no match` here.

**Pull request merged**